File: src/Enterprise/Java/BridgeExample.java
interface Device {
    String powerOn();
    String mute();
}

final class TvDevice implements Device {
    public String powerOn() { return "TV:on"; }
    public String mute() { return "TV:muted"; }
}

final class RadioDevice implements Device {
    public String powerOn() { return "Radio:on"; }
    public String mute() { return "Radio:muted"; }
}

abstract class RemoteControl {
    protected final Device device;
    protected RemoteControl(Device device) { this.device = device; }
    abstract String activate();
}

final class BasicRemote extends RemoteControl {
    BasicRemote(Device device) { super(device); }
    String activate() { return device.powerOn(); }
}

final class MuteRemote extends RemoteControl {
    MuteRemote(Device device) { super(device); }
    String activate() { return device.mute(); }
}

public final class BridgeExample {
    public static void main(String[] args) {
        Device tv = new TvDevice();
        Device radio = new RadioDevice();
        System.out.println("basic-tv=" + new BasicRemote(tv).activate());
        System.out.println("basic-radio=" + new BasicRemote(radio).activate());
        System.out.println("mute-tv=" + new MuteRemote(tv).activate());
        System.out.println("mute-radio=" + new MuteRemote(radio).activate());
    }
}

File: src/Enterprise/Kotlin/BridgeExample.kt
interface Device {
    fun powerOn(): String
    fun mute(): String
}

class TvDevice : Device {
    override fun powerOn() = "TV:on"
    override fun mute() = "TV:muted"
}

class RadioDevice : Device {
    override fun powerOn() = "Radio:on"
    override fun mute() = "Radio:muted"
}

abstract class RemoteControl(protected val device: Device) {
    abstract fun activate(): String
}

class BasicRemote(device: Device) : RemoteControl(device) {
    override fun activate() = device.powerOn()
}

class MuteRemote(device: Device) : RemoteControl(device) {
    override fun activate() = device.mute()
}

fun main() {
    val tv = TvDevice()
    val radio = RadioDevice()
    println("basic-tv=${BasicRemote(tv).activate()}")
    println("basic-radio=${BasicRemote(radio).activate()}")
    println("mute-tv=${MuteRemote(tv).activate()}")
    println("mute-radio=${MuteRemote(radio).activate()}")
}
